pkg/kademlia/kademlia.go

```diff
@@ -5,243 +5,198 @@ package kademlia
-	if start == "" {
-		start = k.dht.GetSelfID()
-	}
-
-	nn, err := k.dht.FindNodes(ctx, start, limit)
-	if err != nil {
-		return []*proto.Node{}, err
-	}
-
-	nodes := convertNetworkNodes(nn)
-
-	for _, r := range restrictions {
-		nodes = restrict(r, nodes)
-	}
-	return nodes, nil
+func (k *Kademlia) GetNodes(ctx context.Context, start string, limit int, restrictions ...proto.Restriction) ([]*proto.Node, error) {
+	// TODO(coyle)
+	return []*proto.Node{}, errors.New("TODO")
 }
 
 // GetRoutingTable provides the routing table for the Kademlia DHT
 func (k *Kademlia) GetRoutingTable(ctx context.Context) (dht.RoutingTable, error) {
-	return &RoutingTable{
-		// ht:  k.dht.HT,
-		// dht: k.dht,
-	}, nil
+	return k.routingTable, nil
 
 }
 
 // Bootstrap contacts one of a set of pre defined trusted nodes on the network and
 // begins populating the local Kademlia node
 func (k *Kademlia) Bootstrap(ctx context.Context) error {
-	return k.dht.Bootstrap()
-}
-
-// Ping checks that the provided node is still accessible on the network
-func (k *Kademlia) Ping(ctx context.Context, node proto.Node) (proto.Node, error) {
-	n, err := convertProtoNode(node)
-	if err != nil {
-		return proto.Node{}, err
+	// What I want to do here is do a normal lookup for myself
+	// so call lookup(ctx, nodeImLookingFor)
+	if len(k.bootstrapNodes) == 0 {
+		return BootstrapErr.New("no bootstrap nodes provided")
 	}
 
-	ok, err := k.dht.Ping(n)
-	if err != nil {
-		return proto.Node{}, err
-	}
-	if !ok {
-		return proto.Node{}, NodeErr.New("node unavailable")
-	}
-	return node, nil
+	_, err := k.lookup(ctx, node.StringToID(k.routingTable.self.GetId()), lookupOpts{amount: 5})
+	return err
 }
 
-// FindNode looks up the provided NodeID first in the local Node, and if it is not found
-// begins searching the network for the NodeID. Returns and error if node was not found
-func (k *Kademlia) FindNode(ctx context.Context, ID dht.NodeID) (proto.Node, error) {
-	nodes, err := k.dht.FindNode(ID.Bytes())
-	if err != nil {
-		return proto.Node{}, err
-
+func (k *Kademlia) lookup(ctx context.Context, target dht.NodeID, opts lookupOpts) (*proto.Node, error) {
+	conc := opts.amount
+	kb := k.routingTable.K()
+	if conc <= 0 {
+		conc = kb
 	}
 
-	for _, v := range nodes {
-		if string(v.ID) == ID.String() {
-			return proto.Node{Id: string(v.ID), Address: &proto.NodeAddress{
-				Transport: defaultTransport,
-				Address:   fmt.Sprintf("%s:%d", v.IP.String(), v.Port),
-			},
-			}, nil
-		}
+	// look in routing table for targetID
+	nodes, err := k.routingTable.FindNear(target, kb)
+	if err != nil {
+		return nil, err
 	}
-	return proto.Node{}, NodeErr.New("node not found")
-}
 
-// ListenAndServe connects the kademlia node to the network and listens for incoming requests
-func (k *Kademlia) ListenAndServe() error {
-	if err := k.dht.CreateSocket(); err != nil {
-		return err
+	// if we have the node in our routing table and that node is not this node (bootstrap) just return the node
```

Comment:
stale comment? also, bootstrap doesn't benefit from this check. it might be worth making a shared function that does the actual lookup that bootstrap and this method both call directly

Reply:
bootstrap doesn't benefit from this true, but it will never matter for bootstrap because the routing table will never return itself if I am calling a find near for self